File: backend/src/main/java/com/travelbilling/entity/Bill.java
package com.travelbilling.entity;

import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "bills")
public class Bill {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String billNumber;
    private Double amount;
    private LocalDateTime billDate;
    private String companyName;
    private String vehicleName;
    private String dutySlipNo;
    private LocalDateTime tripDate;
    private String vehicleType;
    private String acNonAc;
    private Double totalKms;
    private Double totalHours;
    private Double extraKms;
    private Double extraHours;
    private String tripType;
    private Double baseAmount;
    private Double driverBata;
    private Double parking;
    private Double toll;
    private Double nightCharges;
    private Double otherCharges;

    @Column(length = 1000)
    private String notes;

    @Column(columnDefinition = "TEXT")
    private String dynamicCharges;

    private String contactPerson;
    private String bookedBy;
    private String managerName;

    private Double grandTotal;
    private String createdBy;

    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    @ManyToOne
    @JoinColumn(name = "company_id")
    private Company company;

    @ManyToOne
    @JoinColumn(name = "vehicle_id")
    private Vehicle vehicle;

    @OneToMany(mappedBy = "bill")
    private List<Payment> payments;

    @PrePersist
    protected void onCreate() {
        createdAt = updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: backend/app/schemas/bill.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, List
from datetime import date, datetime

class ChargeDTO(BaseModel):
    name: str
    description: Optional[str] = None
    amount: float

class BillRequest(BaseModel):
    billDate: date
    companyName: str
    vehicleName: str
    dutySlipNo: str
    tripDate: Optional[date] = None
    vehicleType: Optional[str] = None
    acNonAc: Optional[str] = None
    totalKms: Optional[float] = 0.0
    totalHours: Optional[float] = 0.0
    extraKms: Optional[float] = 0.0
    extraHours: Optional[float] = 0.0
    tripType: Optional[str] = None
    pricingType: Optional[str] = None
    baseAmount: Optional[float] = 0.0
    driverBata: Optional[float] = 0.0
    parking: Optional[float] = 0.0
    toll: Optional[float] = 0.0
    nightCharges: Optional[float] = 0.0
    otherCharges: Optional[float] = 0.0
    dynamicCharges: Optional[List[ChargeDTO]] = None
    notes: Optional[str] = None
    contactPerson: Optional[str] = None
    bookedBy: Optional[str] = None
    managerName: Optional[str] = None
    rawValues: Optional[str] = None
    originalDoc: Optional[str] = None

class BillResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    billNumber: str
    billDate: Optional[date] = None
    companyName: Optional[str] = None
    vehicleName: Optional[str] = None
    dutySlipNo: Optional[str] = None
    tripDate: Optional[date] = None
    vehicleType: Optional[str] = None
    acNonAc: Optional[str] = None
    totalKms: Optional[float] = 0.0
    totalHours: Optional[float] = 0.0
    extraKms: Optional[float] = 0.0
    extraHours: Optional[float] = 0.0
    tripType: Optional[str] = None
    pricingType: Optional[str] = None
    baseAmount: Optional[float] = 0.0
    driverBata: Optional[float] = 0.0
    parking: Optional[float] = 0.0
    toll: Optional[float] = 0.0
    nightCharges: Optional[float] = 0.0
    otherCharges: Optional[float] = 0.0
    dynamicCharges: Optional[List[ChargeDTO]] = None
    notes: Optional[str] = None
    grandTotal: Optional[float] = 0.0
    createdBy: Optional[str] = None
    createdAt: Optional[datetime] = None
    contactPerson: Optional[str] = None
    bookedBy: Optional[str] = None
    managerName: Optional[str] = None
    rawValues: Optional[str] = None
    originalDoc: Optional[str] = None
